apache/beam
# Retrieve workflow runs for GA Post-Commits Status dashboard

This is a continuation of the #26645.

As part of the previous PR an issue with displaying the GA Post-Commits Status dashboard was fixed.

This PR fixes the data retrieving approach and the following changes have been added:
- Fixed the problem with links to workflows and their run results
- Sequential requests have been changed to simultaneous asynchronous requests
- Added a backoff mechanism with on-the-fly token changes to overcome throttling issues
- Use access token to have authenticated requests to increase the rate limit threshold

The changes are deployed and the dashboard is up-to-date: http://metrics.beam.apache.org/d/CTYdoxP4z/ga-post-commits-status?orgId=1

### How has it been tested?

Tested on the local machine.

1. Follow the [instruction](https://cwiki.apache.org/confluence/display/BEAM/Community+Metrics#CommunityMetrics-GrafanaUI) to deploy and configure a Grafana instance locally
2. Go to the `beamsyncgithub` container: `docker exec -it "beamsyncgithub" bash`
3. Pull out data from GitHub by running the sync script `python ./sync_workflows.py`
4. Go to the `GA Post-Commits Status` Grafana dashboard and verify that data is displayed

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:


See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.
